# FINMA Python Lab 08: NumPy, Pandas and Matplotlib for Financial Analysis

## Overview

In this lab, you'll practice:
- NumPy arrays and operations
- Financial calculations with NumPy
- Pandas DataFrames for tabular data
- Reading and writing data with Pandas
- Data manipulation and aggregation
- Time series analysis
- Calculating financial metrics
- Portfolio analysis

**Important:** Complete your work and have it manually checked by your instructor.



---

## Programming Exercises

### Exercise 1: Portfolio Value Calculation (NumPy)

Using NumPy arrays:
1. Create arrays for: symbols, shares, and prices
2. Calculate the value of each position
3. Calculate the total portfolio value
4. Calculate the percentage allocation for each position

**Data:**
- AAPL: 150 shares at $163.75
- GOOGL: 75 shares at $155.30
- MSFT: 100 shares at $406.20
- AMZN: 60 shares at $200.45
- TSLA: 40 shares at $276.30

In [2]:
# Exercise 1: Portfolio Value Calculation
# Write your code here
import numpy as np
symbols = np.array(['AAPL', 'GOOGL', 'MSFT', 'AMZN', "TSLA"])
shares = np.array([150, 75, 100, 60, 40])
price = np.array([163.75, 155.30, 406.20, 200.45, 276.30])

portfolio_value= shares*price
total_value = np.sum(portfolio_value)
allocation=portfolio_value/total_value*100

print("Portfolio Value per Stock:", portfolio_value)
print("Total Portfolio Value:", total_value)
print("Allocation per Stock:", allocation)

Portfolio Value per Stock: [24562.5 11647.5 40620.  12027.  11052. ]
Total Portfolio Value: 99909.0
Allocation per Stock: [24.58487223 11.65810888 40.65699787 12.03795454 11.06206648]


### Exercise 2: Return Statistics (NumPy)

Given daily returns for a stock, calculate:
1. Average daily return
2. Volatility (standard deviation)
3. Annualized return (assume 252 trading days)
4. Annualized volatility
5. Sharpe ratio (assume risk-free rate = 2%)

**Formula:** Sharpe Ratio = (Annualized Return - Risk Free Rate) / Annualized Volatility

**Data:** Use returns array: `[1.2, -0.5, 0.8, 1.5, -0.3, 0.9, 1.1, -0.7, 1.3, 0.6]`

In [3]:
# Exercise 2: Return Statistics
# Write your code here
daily_returns = np.array([1.2, -0.5, 0.8, 1.5, -0.3, 0.9, 1.1, 1.3, 0.6])
avg_daily_return = np.mean(daily_returns)
print("Average Daily Return:", round(avg_daily_return, 2))

std_dev = np.std(daily_returns)
print("Standard Deviation of Daily Returns:", round(std_dev, 2))

annualised_return = avg_daily_return * 252
annualised_std_dev = std_dev * 252
print("Annualised Return:", round(annualised_return, 2))
print("Annualised Standard Deviation:", round(annualised_std_dev, 2))

sharpe_ratio =(annualised_return - 2)/annualised_std_dev
print("Sharpe Ratio:", round(sharpe_ratio, 2))

Average Daily Return: 0.73
Standard Deviation of Daily Returns: 0.66
Annualised Return: 184.8
Annualised Standard Deviation: 165.89
Sharpe Ratio: 1.1


### Exercise 3: Price Matrix Analysis (NumPy)

Create a 2D NumPy array representing 10 days of prices for 5 stocks.

Calculate:
1. Average price for each stock (across all days)
2. Highest price for each stock
3. Lowest price for each stock
4. Which stock had the highest average price?
5. Which day had the highest average price across all stocks?

**Use the first 10 days from stock_prices_timeseries.csv**

In [5]:
# Exercise 3: Price Matrix Analysis
import csv 
import pandas as pd 
import numpy as np
stock = pd.read_csv("stock_prices_timeseries.csv")
selected_stocks = np.array((stock["AAPL"][:10], stock["GOOGL"][:10], stock["MSFT"][0:10], stock["AMZN"][0:10], stock["TSLA"][0:10]))
stocknames = np.array(["AAPL", "GOOGL", "MSFT", "AMZN", "TSLA"])
dates = stock["Date"][:10].values

highest_avg_price = 0
highest_avg_price_stock = None

for i in range(len(stocknames)):
    avg = np.mean(selected_stocks[i])
    maxprice = np.max(selected_stocks[i])
    minprice = np.min(selected_stocks[i])

    print("the average price for stock", stocknames[i], "is", round(avg, 2))
    print("the maximum price for stock", stocknames[i], "is", round(maxprice, 2))
    print("the minimum price for stock", stocknames[i], "is", round(minprice, 2))
    print()

    if avg > highest_avg_price:
        highest_avg_price = avg
        highest_avg_price_stock = stocknames[i]

print("the stock with the highest avg price is", highest_avg_price_stock)

highest_avg_day = None
highest_avg_day_value = 0
for day in range(len(selected_stocks.T)): 
    dailyavg = np.mean(selected_stocks.T[day])
    #print("Daily average on day", dates[day], "is", dailyavg)

    if dailyavg>highest_avg_day_value: 
        highest_avg_day_value = dailyavg
        highest_avg_day = dates[day]

print("the day with the highest avg price is on", highest_avg_day)


the average price for stock AAPL is 188.87
the maximum price for stock AAPL is 193.6
the minimum price for stock AAPL is 184.25

the average price for stock GOOGL is 2769.85
the maximum price for stock GOOGL is 2812.35
the minimum price for stock GOOGL is 2728.91

the average price for stock MSFT is 382.64
the maximum price for stock MSFT is 392.45
the minimum price for stock MSFT is 372.89

the average price for stock AMZN is 3126.91
the maximum price for stock AMZN is 3175.28
the minimum price for stock AMZN is 3087.56

the average price for stock TSLA is 252.66
the maximum price for stock TSLA is 267.23
the minimum price for stock TSLA is 238.45

the stock with the highest avg price is AMZN
the day with the highest avg price is on 2024-01-12


### Exercise 4: DataFrame Creation and Analysis (Pandas)

Create a DataFrame from the company_info.csv file and:
1. Display the first 5 rows
2. Show summary statistics
3. Find the company with the highest market cap
4. Find the company with the highest P/E ratio
5. Calculate the average dividend yield by sector
6. List all Technology sector companies

In [ ]:
# Exercise 4: DataFrame Creation and Analysis
# Write your code here
import pandas as pd 
import csv

    
df = pd.read_csv("company_info.csv")
print(df.head())
print(df.describe())

#sorting by max market cap
sorteddf = df.sort_values("market_cap", ascending=False)
max_comp = sorteddf.at[0, "company"]
max_marketcap = sorteddf.at[0, "market_cap"]
print("Company with max market cap is", max_comp, "at", max_marketcap )
#one easy way 
#df["market_cap"].idxmax() #returns the index of what we are looking for 
#print(df.loc[df["market_cap"].idxmax()])
#print(df["market_cap"].nlargest(4).index[3])
#f.nlargest(3, 'market_cap').iloc[2]

#highest PE ratio 
highest_pe = df.loc[df["pe_ratio"].idxmax()]
print(highest_pe)

#avg yield per sector
sector_sort = df.sort_values("sector")
#print(sector_sort)
print(df[df["sector"]=="Technology"]["dividend_yield"].mean())
#avg yield per sector - using groupby
sector_yields = df.groupby("sector")["dividend_yield"].mean()
print(sector_yields)




  symbol                company      sector              industry  market_cap  \
0   AAPL             Apple Inc.  Technology  Consumer Electronics      3000.5   
1   MSFT  Microsoft Corporation  Technology              Software      2800.2   
2  GOOGL          Alphabet Inc.  Technology     Internet Services      1750.8   
3   AMZN        Amazon.com Inc.  Technology            E-commerce      1650.3   
4   TSLA             Tesla Inc.  Technology     Electric Vehicles       850.7   

   pe_ratio  dividend_yield  
0      28.5             0.5  
1      32.1             0.7  
2      25.4             0.0  
3      45.2             0.0  
4      75.8             0.0  
        market_cap   pe_ratio  dividend_yield
count    20.000000  20.000000       20.000000
mean    831.020000  26.405000        1.820000
std     839.543477  17.795194        1.511221
min     140.500000  10.200000        0.000000
25%     324.225000  14.525000        0.400000
50%     450.500000  23.500000        2.150000
75%     938

In [92]:
import csv

# Sample data as dictionaries
holdings = [
    {'symbol': 'AAPL', 'shares': 100, 'price': 150.75},
    {'symbol': 'GOOGL', 'shares': 50, 'price': 138.21},
    {'symbol': 'MSFT', 'shares': 75, 'price': 378.91}
]

# Write to CSV
with open('holdings.csv', 'w', newline='') as file:
    fieldnames = ['symbol', 'shares', 'price']
    csv_writer = csv.DictWriter(file, fieldnames=fieldnames)
    
    csv_writer.writeheader()
    csv_writer.writerows(holdings)

print("CSV file written with DictWriter!")

# Read it back
with open('holdings.csv', 'r') as file:
    print(file.read())

CSV file written with DictWriter!
symbol,shares,price
AAPL,100,150.75
GOOGL,50,138.21
MSFT,75,378.91



### Exercise 5: Transaction Analysis (Pandas)

Read transactions.csv and:
1. Calculate total shares bought and sold for each symbol
2. Calculate total commission paid
3. Calculate average buy price and average sell price for each symbol
4. Determine current holdings (shares bought - shares sold)
5. Export a summary to CSV with columns: symbol, shares_held, avg_buy_price, total_commission

In [ ]:
# Exercise 5: Transaction Analysis
# Write your code here
import pandas as pd
trades = pd.read_csv("transactions.csv")


#total shares brought and sold per symbol
avg_prices = trades.groupby(["symbol", "action"]).agg(  {"price": "mean", "shares": "sum"})
#print(avg_prices)

#total commission paid 
total_commission = trades.groupby("symbol")["commission"].sum()
#print(total_commission)

#current holdings
BUYS = trades.loc[trades["action"] =="BUY"].groupby("symbol")["shares"].sum()
SELLS = trades.loc[trades["action"] =="SELL"].groupby("symbol")["shares"].sum()
holdings = BUYS.subtract(SELLS, fill_value=0)
print(holdings)

#printing to csv
avg_buy_price = trades.loc[trades["action"]=="BUY"].groupby("symbol")["price"].mean()
summary = pd.DataFrame({
    "shares_held": holdings,
    "avg_buy_price": avg_buy_price,
    "total_commission": total_commission
})
summary.reset_index().to_csv("holdings.csv")
with open("holdings.csv", "r") as file:
    print(file.read())



TypeError: unsupported operand type(s) for -: 'str' and 'str'

### Exercise 6: Time Series Returns (Pandas)

Using stock_prices_timeseries.csv:
1. Read the data with Date as index
2. Calculate daily returns for all stocks
3. Calculate cumulative returns for all stocks
4. Find which stock had:
   - Highest total return
   - Lowest total return
   - Highest volatility (std of daily returns)
   - Lowest volatility
5. Export the daily returns to CSV

In [ ]:
# Exercise 6: Time Series Returns
# Write your code here
import pandas as pd 
import csv 
import numpy as np

stock_prices = pd.read_csv("stock_prices_timeseries.csv", index_col = "Date")
daily_returns = stock_prices.pct_change()
#print(daily_returns)

cumulative_returns = (1+daily_returns).cumprod()-1
#print(cumulative_returns)

total_returns = daily_returns.sum()
highest_total_return = total_returns.loc[total_returns.idxmax()]
lowest_total_return = total_returns.loc[total_returns.idxmin()]
print("the stock with the highest total return was", (total_returns.idxmax()), "at", highest_total_return)
print("the stock with the lowest total return was", (total_returns.idxmin()), "at", lowest_total_return)

volatility = daily_returns.std()
highest_vol = volatility.loc[volatility.idxmax()]
lowest_vol = volatility.loc[volatility.idxmin()]
print("the stock with the highest volatility was", (volatility.idxmax()), "at", highest_vol)
print("the stock with the lowest volatility was", (volatility.idxmin()), "at", lowest_vol)

daily_returns.to_csv("daily_returns.csv")

the stock with the highest total return was NVDA at 0.3027304336693625
the stock with the lowest total return was GOOGL at 0.056663890950925166
the stock with the highest volatility was NVDA at 0.052740230769191924
the stock with the lowest volatility was GOOGL at 0.009535495601021227


### Exercise 7: Moving Averages (Pandas)

Using stock_prices_timeseries.csv:
1. Calculate 5-day and 10-day moving averages for AAPL
2. Identify days where price crosses above the 5-day MA (potential buy signal)
3. Identify days where price crosses below the 5-day MA (potential sell signal)
4. Calculate how many buy and sell signals occurred
5. Create a new DataFrame with columns: Date, Price, MA5, MA10, Signal

In [ ]:
# Exercise 7: Moving Averages
# Write your code here



### Exercise 8: Portfolio Performance Report (Combined)

Create a comprehensive portfolio performance report:

1. Read transactions.csv to determine current holdings
2. Read stock_prices_timeseries.csv to get latest prices
3. Merge company_info.csv to get sector information
4. Calculate for each holding:
   - Current value
   - Cost basis
   - Unrealized gain/loss
   - Return percentage
5. Calculate portfolio-level metrics:
   - Total value
   - Total cost
   - Total gain/loss
   - Overall return %
   - Allocation by sector
6. Export to CSV with all details

**Hint:** Use the last price in the time series as the current price

In [ ]:
import pandas as pd 
import numpy as np
import csv

transactions = pd.read_csv("transactions.csv")
stock_prices = pd.read_csv("stock_prices_timeseries.csv", index_col = "Date")
company_info = pd.read_csv("company_info.csv")

#current value
current_price=stock_prices.iloc[-1:]
#holdings information
BUY = transactions.loc[transactions["action"]=="BUY"].groupby("symbol")["shares"].sum() #number of shares brought
SELL = transactions.loc[transactions["action"]=="SELL"].groupby("symbol")["shares"].sum() #number of shares sold
net_holdings = BUY.subtract(SELL, fill_value=0) #number of each share once all transactions have occurred

#orginal price information
BUY_DATA = transactions.loc[transactions["action"]=="BUY"]
SELL_DATA = transactions.loc[transactions["action"]=="SELL"]
BUY_COST = BUY_DATA["shares"]*BUY_DATA["price"]+BUY_DATA["commission"]
SELL_COST = SELL_DATA["shares"]*SELL_DATA["price"]-SELL_DATA["commission"]
#holding metrics
TOTAL_BUY_COST = BUY_COST.groupby(BUY_DATA["symbol"]).sum()
AVG_COST_PER_SHARE = TOTAL_BUY_COST/BUY #includes commission!!!
cost_basis=AVG_COST_PER_SHARE*net_holdings #cost basis -> buy only? 
current_value = current_price*net_holdings
unrealised_pl=current_value-cost_basis
return_percentage=(unrealised_pl/cost_basis)*100
print("current_value per holding is", current_value)
print("cost basis per holding is", cost_basis)
print("unrealised profit or loss per holding is", unrealised_pl)
print("return percentage per holding is", return_percentage)

#portfolio level matrics 
portfolio_value = current_value.T.sum()
portfolio_cost = cost_basis.sum()
total_gain = unrealised_pl.T.sum()
overall_return = total_gain/portfolio_cost*100
print(portfolio_value)
print(portfolio_cost)
print(total_gain)
print(overall_return)


summary = pd.DataFrame([{
    "portfolio_value": portfolio_value,
    "portfolio_cost": portfolio_cost,
    "total_gain": total_gain,
    "overall_return_pct": overall_return
}])

summary.to_csv("portfolio_summary.csv", index=False)


Date
2024-01-30    291939.6
dtype: float64
252845.7456944445
Date
2024-01-30    39093.854306
dtype: float64
Date
2024-01-30    15.461543
dtype: float64


In [46]:
import sys
!{sys.executable} -m pip install matplotlib

### Exercise 9: Correlation Analysis (Pandas + NumPy)

Analyze correlations between stocks:
1. Read stock_prices_timeseries.csv
2. Calculate daily returns for all stocks
3. Create a correlation matrix of returns
4. Find the pair of stocks with:
   - Highest correlation
   - Lowest correlation
5. Calculate portfolio variance for an equal-weighted portfolio

**Hint:** Use `df.corr()` for correlation matrix

In [24]:
# Exercise 9: Correlation Analysis
# Write your code here
import pandas as pd
stockprices = pd.read_csv("stock_prices_timeseries.csv", index_col ="Date")
dailyreturns = stockprices.pct_change()
corr_matrix =dailyreturns.corr()
cleaned_matrix = corr_matrix[corr_matrix<1]
cleaned_matrix.idxmax()
cleaned_matrix.idxmin()




AAPL      NVDA
GOOGL     AMZN
MSFT      AMZN
AMZN      NVDA
TSLA        KO
JPM      GOOGL
META      AMZN
NVDA      AMZN
BAC       NVDA
JNJ       NVDA
PFE       MSFT
UNH       AMZN
XOM      GOOGL
CVX      GOOGL
WMT       AMZN
PG        AMZN
KO        AMZN
HD        NVDA
COP      GOOGL
dtype: object

### Exercise 10: Risk-Adjusted Returns (Advanced)

Calculate risk-adjusted metrics for each stock:
1. Calculate daily returns and annualized returns (252 trading days)
2. Calculate volatility (std) and annualized volatility
3. Calculate Sharpe Ratio (risk-free rate = 2%)
4. Calculate Maximum Drawdown for each stock
5. Create a summary DataFrame ranking stocks by:
   - Total return
   - Sharpe ratio
   - Risk (volatility)
6. Export the analysis to CSV

**Formula for Maximum Drawdown:**
```
Running maximum = cumulative maximum of prices
Drawdown = (Current price - Running maximum) / Running maximum
Maximum Drawdown = minimum drawdown value
```

In [25]:
# Exercise 10: Risk-Adjusted Returns
# Write your code here



### Exercise 11: Plot Return Distribution (Matplotlib)

Plot a histogram of provided daily returns:
1. Use the returns array below.
2. Plot a histogram with 20–30 bins, add a vertical line at the mean, and show grid/labels/title.
3. Annotate the chart with mean and standard deviation.
4. Optionally save as `returns_histogram.png`.

**Data:**
- `returns = [0.004, -0.003, 0.006, 0.002, 0.005, -0.001, 0.007, 0.003, -0.002, 0.004, 0.006, 0.001, 0.005, -0.002, 0.003, 0.004, 0.002, 0.005, 0.006, 0.003]`

In [26]:
# Exercise 11: Plot Return Distribution
# Write your code here
import matplotlib.pyplot as plt 
returns = [0.004, -0.003, 0.006, 0.002, 0.005, -0.001, 0.007, 0.003, -0.002, 0.004, 0.006, 0.001, 0.005, -0.002, 0.003, 0.004, 0.002, 0.005, 0.006, 0.003]
meanreturn = sum(returns)/len(returns)
stddev = np.std(np.array(returns))

plt.figure(figsize=[8,8])
plt.hist(returns, bins=25, color="steelblue")
plt.vlines(x=meanreturn, ymin=0, ymax=3, colors='g', linestyles='solid', label='mean return')
plt.text(-0.002, 2.5, f"Mean return is {round(meanreturn,4)}")
plt.text(-0.002, 2.4, f"standard deviation is {round(stddev,4)}")
plt.title("Daily Returns Distribution")
plt.xlabel("Returns")
plt.ylabel("Frequency")
plt.legend()
plt.grid(True)
plt.tight_layout()



NameError: name 'np' is not defined

### Exercise 12: Plot Hypothetical Share Prices (Matplotlib)

Plot provided price paths for two stocks:
1. Use the given dates and prices.
2. Plot both series on one line chart with title, axes labels, legend, and grid.
3. Optionally save as `hypothetical_prices.png`.

**Data:**
- `dates = ["2024-01-02", "2024-01-03", "2024-01-04", "2024-01-05", "2024-01-08", "2024-01-09", "2024-01-10", "2024-01-11", "2024-01-12", "2024-01-15"]`
- `aapl = [150.0, 151.5, 150.8, 152.3, 153.0, 154.2, 153.7, 155.0, 154.5, 156.0]`
- `msft = [320.0, 321.0, 322.5, 321.8, 323.0, 324.5, 325.0, 324.0, 326.5, 327.0]`

In [ ]:
# Exercise 12: Plot Hypothetical Share Prices
# Write your code here


---

## Summary: Key Concepts

### NumPy:
```python
# Array creation
arr = np.array([1, 2, 3])
zeros = np.zeros(5)
ones = np.ones((3, 4))  # 3x4 matrix
range_arr = np.arange(0, 10, 2)

# Operations (vectorized)
arr * 2              # Multiply all elements
arr1 + arr2          # Element-wise addition
np.mean(arr)         # Average
np.std(arr)          # Standard deviation
np.sum(arr)          # Sum

# 2D operations
matrix.mean(axis=0)  # Mean of each column
matrix.mean(axis=1)  # Mean of each row
```

### Pandas:
```python
# Reading data
df = pd.read_csv('file.csv')
df['Date'] = pd.to_datetime(df['Date'])
df.set_index('Date', inplace=True)

# Selection
df['column']                # Single column (Series)
df[['col1', 'col2']]        # Multiple columns
df[df['price'] > 100]       # Filter rows
df.iloc[0]                  # By position
df.loc['2025-12-01']        # By label

# Operations
df['new'] = df['a'] * df['b']  # New column
df.pct_change()                # Percentage change
df.rolling(window=5).mean()    # Moving average
df.groupby('sector').mean()    # Group and aggregate
pd.merge(df1, df2, on='key')   # Merge DataFrames

# Statistics
df.describe()           # Summary statistics
df.mean()              # Column means
df.corr()              # Correlation matrix
```

### Financial Calculations:
```python
# Returns
returns = prices.pct_change()
cum_returns = (1 + returns).cumprod() - 1

# Volatility (annualized)
volatility = returns.std() * np.sqrt(252)

# Sharpe Ratio
sharpe = (annual_return - risk_free) / volatility

# Moving averages
ma = prices.rolling(window=20).mean()
```

---

## Testing and Submission

**Before moving on:**
1. Complete all exercises using the sample data
2. Verify your calculations make sense
3. Export results to CSV where requested
4. Understand the difference between NumPy and Pandas
5. Know when to use each library
6. Have your instructor manually check your work

**Excellent work completing Lab 8!** 🎉